### PR TITLE
Disallow Options source gen produce diagnostics from outside the compilation

### LINKED ISSUE #88152: OptionsValidator source-gen causes VS analyzer to crash
### Description

Recently added OptionsValidator (issue #85475 and PR #87587) under some circumstances emits a diagnostics with location outside of the compilation unit it works on. This causes VS analyzers to crash.

### Reproduction Steps

Source code for the first assembly:
```csharp
using Microsoft.Extensions.Options;

namespace MyAssembly;

[OptionsValidator]
public partial class MyOptionsValidator : IValidateOptions<MyOptions>
{
}

public class MyOptions
{
    [ValidateObjectMembers]
    public AnotherAssembly.ClassInAnotherAssembly? TransitiveProperty { get; set; }
}
```

Source code for another assembly:
```csharp
using System.ComponentModel.DataAnnotations;

namespace AnotherAssembly;

public class ClassInAnotherAssembly
{
    [Required]
    public string? Foo { get; set; }

    // !!! line below causes the generator to emit a warning outside of its compilation !!!
    public SecondClassInAnotherAssembly? TransitiveProperty { get; set; }
}

public class SecondClassInAnotherAssembly
{
    [Required]
    public string? Bar { get; set; }
}
```

### Expected behavior

No errors happen, a warning is emitted normally:
```
SYSLIB1212: Type SecondClassInAnotherAssembly has validation annotations, but member TransitiveProperty doesn't specify [ValidateObjectMembers] which could be an oversight.
```

### Actual behavior

Most of the VS analyzer are crashed:
image

Here's the stacktrace:
```
System.ArgumentException : SyntaxTree is not part of the compilation
Parameter name: syntaxTree
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.GetSemanticModel(SyntaxTree syntaxTree,Boolean ignoreAccessibility)
   at Microsoft.CodeAnalysis.Diagnostics.SuppressMessageAttributeState.IsDiagnosticSuppressed(Diagnostic diagnostic,SuppressMessageInfo& info)
   at Microsoft.CodeAnalysis.Diagnostics.SuppressMessageAttributeState.ApplySourceSuppressions(Diagnostic diagnostic)
   at Microsoft.CodeAnalysis.GeneratorDriver.FilterDiagnostics(Compilation compilation,ImmutableArray`1 generatorDiagnostics,DiagnosticBag driverDiagnostics,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.GeneratorDriver.RunGeneratorsCore(Compilation compilation,DiagnosticBag diagnosticsBag,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.GeneratorDriver.RunGenerators(Compilation compilation,CancellationToken cancellationToken)
   at async Microsoft.CodeAnalysis.SolutionState.CompilationTracker.FinalizeCompilationAsync(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.CodeAnalysis.SolutionState.CompilationTracker.BuildCompilationInfoAsync(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.CodeAnalysis.SolutionState.CompilationTracker.GetOrBuildCompilationInfoAsync(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.CodeAnalysis.SolutionState.CompilationTracker.GetCompilationSlowAsync(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.CodeAnalysis.Document.GetSemanticModelAsync(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.CodeAnalysis.SemanticDocument.CreateAsync(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.CodeAnalysis.IntroduceVariable.AbstractIntroduceVariableService`6.IntroduceVariableAsync[TService,TExpressionSyntax,TTypeSyntax,TTypeDeclarationSyntax,TQueryExpressionSyntax,TNameSyntax](<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.CodeAnalysis.IntroduceVariable.IntroduceVariableCodeRefactoringProvider.ComputeRefactoringsAsync(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.CodeAnalysis.CodeRefactorings.CodeRefactoringService.GetRefactoringFromProviderAsync(<Unknown Parameters>)
```

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

.NET SDK version 8.0.100-preview.7.23327.3

### Other information

_No response_

### PR DESCRIPTION
**Pull request by tarekgh**
Fixes https://github.com/dotnet/runtime/issues/88152

### CODE REVIEW

**Comment by tarekgh** [comment]
@xakep139 @geeknoid could you please have a look?

File: src/libraries/Microsoft.Extensions.Options/gen/Parser.cs [hunk 1]
```diff
@@ -667,12 +667,20 @@ private static string EscapeString(string s)
...
+            {
+                _reportDiagnostic(Diagnostic.Create(desc, location, Array.Empty<object?>()));
+            }
         }
 
         private void Diag(DiagnosticDescriptor desc, Location? location, params object?[]? messageArgs)
         {
-            _reportDiagnostic(Diagnostic.Create(desc, location, messageArgs));
+            // ignore the diagnostics outside the current compilation
+            if (location?.Kind == LocationKind.SourceFile)
```
Comment (xakep139) [inline src/libraries/Microsoft.Extensions.Options/gen/Parser.cs:680]: That covers only half of cases - when an assembly is a 3rd party one. We can still consider producing diagnostics in that case, but the location needs to be some fallback one.

When the assembly is yours (i.e. you have its source code), that condition will be `true`, but VS plugins will crash.

Comment (eiriktsarpalis) [reply src/libraries/Microsoft.Extensions.Options/gen/Parser.cs:680]: I believe the System.Text.Json source generator may be impacted by the same issue (although we haven't received any customer reports about this). What is your recommendation when it comes to picking a fall back location? Just setting to `null`?

cc @layomia who might be interested in this.

Comment (xakep139) [reply src/libraries/Microsoft.Extensions.Options/gen/Parser.cs:680]: We can consider making an `OptionsValidator` to be a fallback in that case (worst case scenario is when a validator is applied on a type residing in another assembly).
For JSON source-gen, a class annotated with `[JsonSerializable]` can be considered as a fallback location.

Comment (xakep139) [reply src/libraries/Microsoft.Extensions.Options/gen/Parser.cs:680]: In both cases types annotated with an attribute that kicks in source-generation should definitely belong to the compilation unit in which a source-gen is executing.

Comment (eiriktsarpalis) [reply src/libraries/Microsoft.Extensions.Options/gen/Parser.cs:680]: Thanks, I'll run an audit of the implementation to see whether fallback is necessary.

Comment (tarekgh) [reply src/libraries/Microsoft.Extensions.Options/gen/Parser.cs:680]: > When the assembly is yours (i.e. you have its source code), that condition will be true, but VS plugins will crash.

Are we sure of that? the compilation will be referencing your assembly as a meta data at that time. no?

Comment (xakep139) [reply src/libraries/Microsoft.Extensions.Options/gen/Parser.cs:680]: I was debugging another generator locally and it was the case for me. You can check your fix by running it on the code I provided in #88152.

**Comment by tarekgh** [comment]
@xakep139 I have modified the fix to let the diagnostics use the closest location of the failure inside the compilation. I tested it with Visual Studio too :-) please have a look. Thanks!

File: src/libraries/Microsoft.Extensions.Options/tests/SourceGeneration.Unit.Tests/Main.cs [hunk 2]
```diff
@@ -1253,8 +1253,12 @@ public class MyOptions
                     new List<string> { source1 })
                 .ConfigureAwait(false);
 
-            Assert.Empty(diagnostics); // no diagnostics should be produced
+            Assert.Equal(1, diagnostics.Count());
```
Comment (xakep139) [inline src/libraries/Microsoft.Extensions.Options/tests/SourceGeneration.Unit.Tests/Main.cs:1256]: `var diag = Assert.Single(diagnostics);`

File: src/libraries/Microsoft.Extensions.Options/tests/SourceGeneration.Unit.Tests/Main.cs [hunk 3]
```diff
@@ -1253,8 +1253,12 @@ public class MyOptions
...
+            Assert.Equal(DiagDescriptors.PotentiallyMissingTransitiveValidation.Id, diagnostics[0].Id);
```
Comment (xakep139) [inline src/libraries/Microsoft.Extensions.Options/tests/SourceGeneration.Unit.Tests/Main.cs:1257]: `Assert.Equal(DiagDescriptors.PotentiallyMissingTransitiveValidation.Id, diag.Id);`

File: src/libraries/Microsoft.Extensions.Options/tests/SourceGeneration.Unit.Tests/Main.cs [hunk 4]
```diff
@@ -1253,8 +1253,12 @@ public class MyOptions
...
             _ = Assert.Single(generatedSources);
+
+            // validate the location is inside the MyOptions class and not outside the compilation which is in the referenced assembly
+            Assert.StartsWith("src-0.cs: (12,", diagnostics[0].Location.GetLineSpan().ToString());
```
Comment (xakep139) [inline src/libraries/Microsoft.Extensions.Options/tests/SourceGeneration.Unit.Tests/Main.cs:1261]: `Assert.StartsWith("src-0.cs: (12,", diag.Location.GetLineSpan().ToString());`

**Review by xakep139 — APPROVED** [review]
Added a few nit comments, but the PR LGTM

**Pull request merged**